fix gentim2d xx and adxx [start,end,diff] rec and file length (#664)
MITgcm/MITgcm: doc/ocean_state_est/ocean_state_est.rst

CTRL: Model Parameter Adjustment Capability
-------------------------------------------

> Author: Gael Forget, An T. Nguyen

Author: i've added Martin's name.

> Author: Gael Forget, An T. Nguyen

Author: I can add this, once we also figure out what needs to be done with the 1a,b vs 1.1 below...

> Author: Gael Forget, An T. Nguyen

Author: I've added Martin's name.

be mindful of control parameter combinations that make sense according to their
specific setup, e.g., with the :ref:`EXF package <ssub_phys_pkg_exf_config>`.

.. _gen_ctrl_rec:

Generic Control Record Access
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
For each control variable ``$ctrlvar``, three pairs of .data files (and their 
corresponding .meta) are required or produced per adjoint run: 

> (1a) ``xx_$ctrlvar.$iternumber.data/meta``,

Reviewer: `$ctrlvar` already contains the `xx_` according to the definition in the beginning of the section.

Author: now fixed.

(1b) ``adxx_$ctrlvar.$iternumber.data/meta``,
(2a) ``xx_$ctrlvar.tmp.$iternumber.data/meta``,
(2b) ``adxx_$ctrlvar.tmp.$iternumber.data/meta``, and
(3a) ``xx_$ctrlvar.effective.$iternumber.data/meta``,
(3b) ``adxx_ad$ctrlvar.effective.$iternumber.data/meta``.

> In an adjoint run, if a previously produced adjustment ``xx_$ctrlvar.$iternumber.data``

Reviewer: even for `optimcycle=0`, this file is generated.

Author: I think the difference between optimcycle=0 and optimcycle>0 is that for the former case, the $ctrlvar.$iternumber.data/meta are also created at this same time, and so we never have to worry size of files, while in the latter case the $ctrlvar.$iternumber.data/meta already exist and now we need to figure out which records from that file is being read and what the size of the effective file is compared to the existing one.

Author: @mjlosch , is this something we still need to address? or can i mark as resolved?

exists for the 2D time-dependent controls, three variables [start,end,diff]rec will be
initialized as a function of the startdate in data.cal and pickup time niter0
in call to packages_init_fixed.F (which calls ctrl_init, ctrl_init_get_gen).
These three variables are subsequently used to determine the length of the three
pairs (1--3) of files above in the order as follows:  

- First the ``adxx_$ctrlvar.[,tmp,effective].$iternumber.`` files (1b,2b,3b above
  are produced in packages_init_fixed.F-->ctrl_init.F-->ctrl_init_ctrlvar
  (with yadprefix="ad") and of sizes 1:endrec for 1b, 1:diffrec for 2b and 3b --
  see :numref:`adxx_creation`; 

> - Second, records startrec:endrec of (1a) ``xx_$ctrvar.$iternumber.data`` are

Reviewer: The way I read `S/R ctrl_get_gen`, it *only* reads `$ctrlvar.effective.$iternumber.data`` is accessed during the simulation, unless `optimcycle<0` (does that ever happen?).

Author: now fixed.

Author: I'm not sure i understand your question.. is there something missing in the sentence?  I don't think you ever have a case with optimcycle < 0, not that i'm aware of!

Reviewer: Sorry, edited the question, but I meant this: what is described in this text does not correspond to what I see in `S/R ctrl_get_gen`, where for `optimcycle .ge. 0` the file to be read is always the "effective" version `xx_$ctrlvar.effective.data` . The text suggests (to me) it is `xx_$ctrvar.$iternumber.data`. But that's maybe just my personal taste, because I don't see where `S/R ctrl_get_gen_rec` accesses anything, but it just compute the record number to be accessed, right?

Author: I believe you might be looking into a different file of the flow?  I've just read through my build directory, this is the flow I see, which is consistent with what I wrote in the documentation (which is that the file xx_$ctrlvar.$iternumber.data is the file that's being read in, not the effective one):

the_model_main.f --> call mdthe_main_loop

inside ad_taf_output.f
mdthe_main_loop -> initialise_variamd -> packages_init_variablesmd -> ctrl_init_variablesmd --> ctrl_map_ini_gentim2dmd --> see attached pix

line 56398-56405: create files , where fnamegenin is the xx_$ctrvar.$iternumber.data , and fnamegenout is the xx_$ctrvar.effective.$iternumber.data

line 56433: reads in from fnamegenin <-- xx_$ctrvar.$iternumber.data
line 56435: writes out to fnamegenout <-- xx_$ctrvar.effective.$iternumber.data

![ctrl_map_ini_gentim2dmd](https://user-images.githubusercontent.com/37635388/223060969-0f5df0e7-1aa6-4adb-a66f-f7219ecdd810.jpg)

Reviewer: I have a suggestion for this, which makes it a little clearer to me. I would avoid using the "md" files in the description directly, they are not different from the original, except that they include code to store fields. I think because of `ctrl_gen_get_rec`, it's not clear where this happens. I add the suggestion to the code.

  initialize_variamd (inside call adthe_main_loop in taf_ad_output.F) --
  see :numref:`xx_creation`. 

  .. figure:: figs/adxx_creation.*
    :width: 95%
    :align: center
    :alt: adxx_creation
    :name: adxx_creation

    adxx* are created first inside packages_init_fixed / ctrl_init.F

The difference in length of records for 1[a,b] compared to 2[a,b] and 3[a,b] is due to
the fact that we need to access records startrec:endrec in 1a, i.e., file 1a needs
a total of at least endrec records, and the file 1b is automatically generated by
taf to require access to endrec:-1:startrec.  File 1b, in particular, is where adjoint
sensitivity will be accumulated backward and written to.  For pairs 2[a,b] and 3[a,b],
because they are generated *after* we have already accessed the correct records 
startrec:endrec in 1a, we simply create / write out these records in a shorter file
of size 1:diffrec = 1:endrec-startrec+1.

After their file size initializations, 3a which has physical unit is passed on to
pkg/exf for surface forcing application.

  .. figure:: figs/xx_creation.*
    :width: 95%
    :align: center
    :alt: xx_creation
    :name: xx_creation

>     xx* are read in/created afterward inside packages_init_variables[md] / ctrl_map_ini_gentim2d[md].   During the adjoint run, adxx* are read from ctrl_map_ini_gentim2d_ad, requesting records in reverse, and can cause crashing if that last rec is difrec instead of endrec.  As a result, file 1b needs accessible records endrec:-1:startrec and the length of the file is endrec.

Reviewer: use `$ctrlvar` and `ad$ctrlvar` instead of "xx*/adxx*", find `diffrec`.

Author: done regarding $ctrlvar and ad$ctrlvar.  I didn't understand the "diffrec" inclusion in the comment, what do you suggest here?

be mindful of control parameter combinations that make sense according to their
specific setup, e.g., with the :ref:`EXF package <ssub_phys_pkg_exf_config>`.

.. _gen_ctrl_rec:

Generic Control Record Access
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
For each control variable ``$ctrlvar``, three pairs of .data files (and their
corresponding .meta) are required or produced per adjoint run:

#. (a) ``$ctrlvar.$iternumber.data/meta``
   (b) ``ad$ctrlvar.$iternumber.data/meta``
#. (a) ``$ctrlvar.tmp.$iternumber.data/meta``
   (b) ``ad$ctrlvar.tmp.$iternumber.data/meta``
#. (a) ``$ctrlvar.effective.$iternumber.data/meta``

>    (b) ``ad$ctrlvar.effective.$iternumber.data/meta``

Author: when i build it in my local dir on my mac, i get this, which seems to show 1. a,b , 2.a,b, 3.a,b:

<img width="388" alt="image" src="https://user-images.githubusercontent.com/37635388/219538937-c2278d2b-b8ed-438f-bf41-ab044aa349a6.png">

>    (b) ``ad$ctrlvar.effective.$iternumber.data/meta``

Reviewer: We can just hardwire this as numbers. I can do this if you like:
```
1. (a) ``$ctrlvar.$iternumber.data/meta``
   (b) ``ad$ctrlvar.$iternumber.data/meta``
2. (a) ``$ctrlvar.tmp.$iternumber.data/meta``
   (b) ``ad$ctrlvar.tmp.$iternumber.data/meta``
3. (a) ``$ctrlvar.effective.$iternumber.data/meta``
   (b) ``ad$ctrlvar.effective.$iternumber.data/meta``
```
The reason for the list was, that I thought it would be easier to read.

>    (b) ``ad$ctrlvar.effective.$iternumber.data/meta``

Reviewer 2: confirming... on my mac it renders ok but readthedocs has it as numbers. Suggest we go with @mjlosch suggestion here.

be mindful of control parameter combinations that make sense according to their
specific setup, e.g., with the :ref:`EXF package <ssub_phys_pkg_exf_config>`.

.. _gen_ctrl_rec:

Generic Control Record Access
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
For each control variable ``$ctrlvar``, three pairs of .data files (and their
corresponding .meta) are required or produced per adjoint run:

::

   1a   $ctrlvar.$iternumber.data/meta
   1b ad$ctrlvar.$iternumber.data/meta

   2a   $ctrlvar.tmp.$iternumber.data/meta
   2b ad$ctrlvar.tmp.$iternumber.data/meta

   3a   $ctrlvar.effective.$iternumber.data/meta
   3b ad$ctrlvar.effective.$iternumber.data/meta

In an adjoint run with the 2D time-dependent controls (CPP-flag
:varlink:`ALLOW_GENTIM2D_CONTROL` defined), three variables

> :varlink:`startrec`, :varlink:`endrec`, and :varlink:`diffrec` will be

Reviewer: ```suggestion
:varlink:`startrec`, :varlink:`endrec`, and :varlink:`diffrec` = :varlink:`endrec` - :varlink:`startrec` will be
```

Author: done.

initialized as a function of the startdate (:varlink:`startdate_1`,
:varlink:`startdate_2`) in data.cal, the control variables startdates
(:varlink:`xx_gentim2d_startdate1`, :varlink:`xx_gentim2d_startdate2`) in
data.ctrl, and the pickup time ``niter0`` in packages_init_fixed.F (which
calls ctrl_init, ctrl_init_rec).  These three variables are subsequently
used to determine the length of the three pairs (1--3) of files above in the
order as follows:

> - First the ``ad$ctrlvar.[,tmp,effective].$iternumber.`` files (1b,2b,3b above
>   are produced in packages_init_fixed.F-->ctrl_init.F-->ctrl_init_ctrlvar
>   (with yadprefix="ad") and of sizes 1:endrec for 1b, 1:diffrec for 2b and 3b --
>   see :numref:`adxx_creation`;
> 
> - Second, records startrec:endrec of (1a) ``$ctrvar.$iternumber.data`` are
>   accessed in ctrl_get_gen_rec.F
>   and (2a,3a) ``$ctrlvar.{tmp,effective}.data`` of size 1:diffrec are produced in
>   initialize_variamd (inside call adthe_main_loop in taf_ad_output.F) --
>   see :numref:`xx_creation`.
> 

Reviewer: ```suggestion
- First the ``ad$ctrlvar.[,tmp,effective].$iternumber.`` files (1b,2b,3b above)
  are initialised with zeros in packages_init_fixed.F-->ctrl_init.F-->ctrl_init_ctrlvar
  (with yadprefix="ad"); 1b has the size 1:endrec and 2b and 3b have the size 1:diffrec --
  see :numref:`adxx_creation`;

- Second, records startrec:endrec of (1a) ``$ctrvar.$iternumber.data`` are
  read in filelink:`ctrl_map_ini_gentim2d.F` and then written to
  (2a,3a) ``$ctrlvar.{tmp,effective}.data`` of size 1:diffrec; note that the actual routines 
  are often the "md" versions produced by taf, e.g. `s/r ctrl_map_ini_gentim2dmd` called from 
  `s/r   initialize_variamd` (called from adthe_main_loop in taf_ad_output.F) 
  --  see :numref:`xx_creation`.

```

Author: Thanks @mjlosch ! i changed your suggested wording a tiny bit , to add in the processing of the records after being read in the case smooth or scaling or whatever needs to be applied, before those records are being written out. also change .F to .f .

            enddo

The difference in length of records for 3[a,b] compared to 1[a,b] and 2[a,b] is
due to the fact that we need to access records
:varlink:`startrec` thru :varlink:`endrec` in 3a, i.e., file 3a needs a total of at
least :varlink:`endrec` records; file 3b is automatically generated to
provide access to :varlink:`endrec` thru :varlink:`startrec` (i.e., in reverse order). File 3b, in
particular, is where adjoint sensitivity will be accumulated backward and
written; note the model would thus crash if its last record were :varlink:`diffrec` rather than :varlink:`endrec`.
For pairs 1[a,b] and 2[a,b], because they are generated *after* we
have already accessed the correct records :varlink:`startrec` to :varlink:`endrec` in 3a, we simply
create and write out these records in the shorter file size :varlink:`diffrec`.
After their file size initializations, the control adjustment field with physical unit from file 1a is passed on
to :filelink:`pkg/exf` for surface forcing application.

Note, that :varlink:`xx_gentim2d_startdate` can be used to control how many
records the different :varlink:`xx_gentim2d` files
contain. :numref:`xx_var_sketch` illustrates a few examples.

>   .. figure:: figs/ctrl_var_sketch.pdf

Reviewer 3: As predicted by @mjlosch, the pdf file does not show up in html and is replaced by a link (at least in the PR build of the docs).  Now in a perfect world, we would be able to use the imgconverter extension to convert any pdfs to png automatically (and keep these large png files out of the repo).  Unfortunately, this is broken.  We could consider providing our own extensions that uses poppler to convert.  For readthedocs, we just have to add an apt_packages option to the readthedocs yml, for local builds, users would have to install poppler-utils.  I've tried this [here](https://github.com/jahn/MITgcm/tree/imgconv), see the result [here](https://mitgcm-jahn.readthedocs.io/en/imgconv/ocean_state_est/ocean_state_est.html#xx-var-sketch).

Reviewer 2: I'm confused here... there are other pdf figures in the doc (the doc in ch 5 even **suggests** using pdfs), and it renders fine for me and seems ok on readthedocs. Did something change?

Reviewer 3: I think the trick is to use filename.* in the image directive and have a png version in the same place.   To avoid the png file, one could have an svg and a pdf file, such as [this](https://mitgcm.readthedocs.io/en/latest/_images/mitgcm_goals.svg).  With an extension, one could avoid having duplicate files.

Reviewer 3: In firefox at least, this doesn't render: https://mitgcm--664.org.readthedocs.build/en/664/ocean_state_est/ocean_state_est.html#xx-var-sketch

Reviewer 2: ok, right, it does say in the doc to include both a pdf and a svg for different renderings. That's what's done for many figures in the manual. Here, I was not able to do a true svg, and when I skipped it, discovered that safari is smart enough to deal with it (but Firefox isn't). So the ch5 doc really should say to include a pdf and a svg (or png) version.

The aside issues here:
- these figures use a lot a space and for multiple reasons we want to minimize, so having to include a bulky png here is not ideal.
- and this is partly my fault, we don't have editable versions of many figures, as I never collected the them in original format (be it some kind of vector graphic or bitmap or whatever). Some I never had in the first place, some are scattered all over my hard drives. Again the problem was we obviously didn't want to include these in the MITgcm repo (why we never made a separate repo, dunno).

Reviewer 2: I'm able to make a svg file from the pdf I generate from the keynote file. The svg is editable in vector graphics form (although not quite as easy/flexible as editing in native keynote). And the svg is <1/2 the size of the png. If we have settled on the final figure I can push new pdf, sgv versions to this PR.

           `|`-:filelink:`ctrl_set_globfld_xy <pkg/ctrl/ctrl_set_globfld_xy.F>`\ (fname(2)) (with yadprefix='ad') 
              `|`-:filelink:`mds_write_field <pkg/mdsio/mdsio_write_field.F>`\ (adxx_atemp.tmp.0000000001)        **<- size diffrec**
       `|`-:filelink:`ctrl_init_ctrlvar <pkg/ctrl/ctrl_init_ctrlvar.F>`\ ('xx_atemp.0000000001)
           `|`-:filelink:`ctrl_set_fname <pkg/ctrl/ctrl_set_fname.F>`\(xx_fname,fname)         **--> fname(1:3)=[,ad,hn]xx_atemp.0000000001**
           `|`-:filelink:`ctrl_set_globfld_xy <pkg/ctrl/ctrl_set_globfld_xy.F>`\ (fname(2)) (with yadprefix='ad') 
              `|`-:filelink:`mds_write_field <pkg/mdsio/mdsio_write_field.F>`\ (adxx_atemp.0000000001)            **<- size endrec**

- Second, within ``initiase_variamd.f`` (see below), records :varlink:`startrec`
  to :varlink:`endrec` of file 3a ``$ctrvar.$iternumber.data`` are read
  in :filelink:`ctrl_map_ini_gentim2d.F <pkg/ctrl/ctrl_map_ini_gentim2d.F>`,
  processed if scaling or smoothing, etc., need to be applied, and then written
  to (1a,2a) ``$ctrlvar.{effective,tmp}.data`` of size :varlink:`diffrec`.

>   Note these routines contain a ``md`` or ``ad`` suffix and are produced by TAF,
>   e.g., ``s/r ctrl_map_ini_gentim2dmd`` called from ``s/r initialize_variamd``
>   (in turn called from ``s/r adthe_main_loop`` in TAF-generated file ``taf_ad_output.f``). 
> 
> .. parsed-literal ::
> 
>   \|-adthe_main_loop  **only available in ad_taf_output.f, called from the_model_main.f**
>     \|-adopen (many tapes, ocean variables, atmos, obcs, etc)  **initialize tapelev grid, etc.**

Reviewer: Since I really recommend using the `-nocat4ad` option when generating the makefile, so that files are not longer concatenated and sent as a single monstrous file, I would like to reduce the number of mentions of this `ad_taf_output.f` file. At least also mention the other file(s) in question, and later (in a different PR) we can refer to section 7.2.3, which would need to include this information as well:
```suggestion
  Note these routines contain a ``md`` or ``ad`` suffix and are produced by TAF,
  e.g., ``s/r ctrl_map_ini_gentim2dmd`` (found in TAF-generated file ``ctrl_map_ini_gentim2d_ad.f``) called from ``s/r initialize_variamd`` (found in TAF-generated file ``initialize_varia_ad.f``)
  (in turn called from ``s/r adthe_main_loop`` in TAF-generated file ``the_main_loop_ad.f``, alternatively all of these routine are found the concatenated file ``ad_taf_output.f``). 

.. parsed-literal ::

  \|-adthe_main_loop  **only available in the_main_loop_ad.f, called from the_model_main.f**
    \|-adopen (many tapes, ocean variables, atmos, obcs, etc)  **initialize tapelev grid, etc.**
```

    \|-initialise_variamd
      \|-packages_init_variablesmd
        \|-:filelink:`diagnostics_init_varia <pkg/diagnostics/diagnostics_init_varia.F>`, :filelink:`kpp_init_varia <pkg/kpp/kpp_init_varia.F>`, :filelink:`exf_init_varia <pkg/exf/exf_init_varia.F>`  **store salt,theta**
        \|-:filelink:`profiles_init_varia <pkg/profiles/profiles_init_varia.F>`, :filelink:`ecco_init_varia <pkg/ecco/ecco_init_varia.F>`, :filelink:`obcs_init_variables <pkg/obcs/obcs_init_variables.F>`  **some done after ctrl**
        \|-ctrl_init_variablesmd
          \|-:filelink:`ctrl_map_ini_genarr <pkg/ctrl/ctrl_map_ini_genarr.F>`
            \|-:filelink:`ctrl_map_genarr2d <pkg/ctrl/ctrl_map_genarr.F>`  **e.g., set etan,siheff ctrl**
            \|-:filelink:`ctrl_map_genarr3d <pkg/ctrl/ctrl_map_genarr.F>`  **e.g., set logdiffkr ctrl**
          \|-ctrl_map_ini_gentim2dmd

>             \|-:filelink:`ctrl_init_rec <pkg/ctrl/ctrl_init_rec.F>`\ (xx_atemp) **example here for atemp: [startrec,endrec,diffrec]=[24,37,14]**
>             \|-:filelink:`active_read_xy <pkg/autodiff/active_file.F>`\ (fnamegenIn,lrec) **read in xx_atemp.0000000001.data from 24->37**
>             \|-:filelink:`active_write_xy <pkg/autodiff/active_file.F>`\ (fnamegenOut,irec) **write out to xx_atemp.effective.0000000001.data from 1->14**
>             \|-:filelink:`active_read_xy <pkg/autodiff/active_file.F>`\ (fnamegenOut,irec)  **read in xx_atemp.effective.0000000001.data 1->14, do some math**
>             \|-:filelink:`active_write_xy <pkg/autodiff/active_file.F>`\ (fnamegenTmp,irec)  **write out to xx_atemp.tmp.0000000001.data 1->14**
>             do irec=1,diffrec
>             \|-:filelink:`active_read_xy <pkg/autodiff/active_file.F>`\ (fnamegenOut,irec)
>             \|-:filelink:`mds_read_field <pkg/mdsio/mdsio_read_field.F>`\ (xx_gentim2d_weight,jrec) **if variaweight, jrec=lrec, else jrec=1**
>             \|-smooth_correl2d  **or smooth2d**
>             \|-xx_gen/sqrt(wgentim2d)  **if doscaling**
>             \|-exch_xy_rl
>             \|-:filelink:`active_write_xy <pkg/autodiff/active_file.F>`\ (fnamegenOut,irec) **write out to xx_atemp.effective.0000000001.data (smooth/scaled)**

Reviewer: I think this is clear, again I suggest to break the lines, so that we don't have scroll sideways and also to make the text "less compact" and easier to follow(?)
```suggestion
            \|-:filelink:`ctrl_init_rec <pkg/ctrl/ctrl_init_rec.F>`\ (xx_atemp)
	             **example here for atemp: [startrec,endrec,diffrec]=[24,37,14]**
            \|-:filelink:`active_read_xy <pkg/autodiff/active_file.F>`\ (fnamegenIn,lrec)
	             **read in xx_atemp.0000000001.data from 24->37**
            \|-:filelink:`active_write_xy <pkg/autodiff/active_file.F>`\ (fnamegenOut,irec)
	             **write out to xx_atemp.effective.0000000001.data from 1->14**
            \|-:filelink:`active_read_xy <pkg/autodiff/active_file.F>`\ (fnamegenOut,irec)
	             **read in xx_atemp.effective.0000000001.data 1->14, do some math**
            \|-:filelink:`active_write_xy <pkg/autodiff/active_file.F>`\ (fnamegenTmp,irec)
	             **write out to xx_atemp.tmp.0000000001.data 1->14**
            do irec=1,diffrec
            \|-:filelink:`active_read_xy <pkg/autodiff/active_file.F>`\ (fnamegenOut,irec)
            \|-:filelink:`mds_read_field <pkg/mdsio/mdsio_read_field.F>`\ (xx_gentim2d_weight,jrec)
	             **if variaweight, jrec=lrec, else jrec=1**
            \|-smooth_correl2d  **or smooth2d**
            \|-xx_gen/sqrt(wgentim2d)  **if doscaling**
            \|-exch_xy_rl
            \|-:filelink:`active_write_xy <pkg/autodiff/active_file.F>`\ (fnamegenOut,irec)
	             **write out to xx_atemp.effective.0000000001.data (smooth/scaled)**
```

::

   1a   $ctrlvar.$iternumber.data
   1b ad$ctrlvar.$iternumber.data

   2a   $ctrlvar.tmp.$iternumber.data
   2b ad$ctrlvar.tmp.$iternumber.data

   3a   $ctrlvar.effective.$iternumber.data

>    3b ad$ctrlvar.effective.$iternumber.data

Author: In looking at the actual code, I would propose we reorder this, so that it reads exactly the way it is in the code (which loops from 1->3 starting with 1=effective, 2=tmp, 3='', so swapping 3->1 and 1->3):

1a     $ctrlvar.effective.$iternumber.data
1b ad$ctrlvar.effective.$iternumber.data

2a (same as in doc now)
2b (same as in doc now)

3a      $ctrlvar.$iternumber.data
3b ad$ctrlvar.$iternumber.data

Reviewer 2: yes, for two reasons. First, the slash is potentially confusing because it looks like a subdirectory. Secondly, in the sentence above it yellow box it says the .meta file is produced for these files, as per usual, so it did not seem necessary to remake the point in the yellow box. (and, I think users' expectation already  is that a meta file is produced with any output .data file dumped by the model)

In an adjoint run with the 2-D time-dependent controls (CPP-flag
:varlink:`ALLOW_GENTIM2D_CONTROL` defined), three variables
:varlink:`startrec`, :varlink:`endrec`, and :varlink:`diffrec` =
:varlink:`endrec` - :varlink:`startrec` + 1 will be
initialized as a function of the startdate (:varlink:`startdate_1`,
:varlink:`startdate_2`) in ``data.cal``, the control variables startdates
(:varlink:`xx_gentim2d_startdate1`, :varlink:`xx_gentim2d_startdate2`) in
``data.ctrl``, and the pickup time :varlink:`niter0` in
:filelink:`packages_init_fixed.F <model/src/packages_init_fixed.F>` (which
calls :filelink:`ctrl_init.F <pkg/ctrl/ctrl_init.F>`,
:filelink:`ctrl_init_rec.F <pkg/ctrl/ctrl_init_rec.F>`). These three variables
are subsequently used to determine the record length of the three pairs (1--3)
of the above files, in the order as follows:

> - First the ``ad$ctrlvar.[,tmp,effective].$iternumber.`` files (1b,2b,3b) above

Author: - First the ``ad$ctrlvar.[effective,tmp,].$iternumber.`` files (1b,2b,3b) above

  are initialized with zeros in
  :filelink:`packages_init_fixed.F <model/src/packages_init_fixed.F>`-->
  :filelink:`ctrl_init.F <pkg/ctrl/ctrl_init.F>`-->
  :filelink:`ctrl_init_ctrlvar.F <pkg/ctrl/ctrl_init_ctrlvar.F>`
  (with :varlink:`yadprefix` = ``'ad'``); 1b has size :varlink:`endrec`

>   and 2b and 3b have size :varlink:`diffrec` -- see :numref:`adxx_creation`;

Author: 1b and 3b have size :varlink:`diffrec` and 3b has size :varlink:`endrec` -- see :numref:`adxx_creation``;

::

>   Flow of pkg/ctrl when the adjoint is running:

Author: Flow of pkg/ctrl when the adjoint is running, assuming for $iternumber=0000000001:

  
  Note: the_model_main.F calls the_main_loop, but once the code is generated from TAF,
  the preprocessed form the_model_main.f calls either mdthe_main_loop or adthe_main_loop
       
  the_model_main.f
  |-initialize_fixed
    |-ini_parms
    |-packages_boot, packages_readparms
    |-set_parms, ini_model_io, ini_grid, load_ref_files, ini_eos, set_ref_state,
      set_grid_factors, ini_depths, ini_masks, etc.
  
    |-packages_init_fixed
      |-cal_init_fixed, diagnostics_init_early, diagnostics_main_init, gad_init_fixed,
        mom_init_fixed, obcs_init_fixed, exf_init_fixed, kpp_init_fixed, gmredi_init_fixed,
        seaice_cost_init_fixed, smooth_init_fixed, ecco_cost_init_fixed, 
        profiles_init_fixed, seaice_init_fixed, salt_plume_init_fixed
  
      |-ctrl_init
       |-active_write_xyz('wunit')
       |-ctrl_init_ctrlvar (genarr2d, genarr3d)

       |-ctrl_init_rec(gentim2d_startdate, diffrec, startrec, endrec)
       |-ctrl_init_ctrlvar('xx_atemp.effective.0000000001, 'c','xy')
           |-ctrl_set_fname(xx_fname,fname)           --> fname(1:3)=[,ad,hn]xx_atemp.effective.0000000001
           |-ctrl_set_globfld_xy(fname(2)) (with yadprefix='ad') 

>               |-MDSWRITEFIELD(adxx_atemp.effective.0000000001)

Author: add (size diffrec) to the right of this line to indicate the size?

       |-ctrl_init_ctrlvar('xx_atemp.tmp.0000000001)
           |-ctrl_set_fname(xx_fname,fname)           --> fname(1:3)=[,ad,hn]xx_atemp.tmp.0000000001
           |-ctrl_set_globfld_xy(fname(2)) (with yadprefix='ad') 

>               |-MDSWRITEFIELD(adxx_atemp.tmp.0000000001)

Author: (size diffrec)

       |-ctrl_init_ctrlvar('xx_atemp.0000000001)
           |-ctrl_set_fname(xx_fname,fname)           --> fname(1:3)=[,ad,hn]xx_atemp.0000000001
           |-ctrl_set_globfld_xy(fname(2)) (with yadprefix='ad') 

>               |-MDSWRITEFIELD(adxx_atemp.0000000001)

Author: (size endrec)

    | :math:`\phantom{xx}` set_grid_factors, ini_depths, ini_masks, etc.

    | :math:`\phantom{x}` `|`-packages_init_fixed
    | :math:`\phantom{xx}` `|`-cal_init_fixed, diagnostics_init_early, diagnostics_main_init, gad_init_fixed,
    | :math:`\phantom{xxx}` mom_init_fixed, obcs_init_fixed, exf_init_fixed, kpp_init_fixed, gmredi_init_fixed,
    | :math:`\phantom{xxx}` seaice_cost_init_fixed, smooth_init_fixed, ecco_cost_init_fixed, profiles_init_fixed,
    | :math:`\phantom{xxx}` seaice_init_fixed, salt_plume_init_fixed

    | :math:`\phantom{xx}` `|`-ctrl_init
    | :math:`\phantom{xxx}` `|`-active_write_xyz('wunit')
    | :math:`\phantom{xxx}` `|`-ctrl_init_ctrlvar (genarr2d, genarr3d)

    | :math:`\phantom{xxx}` `|`-ctrl_init_rec(gentim2d_startdate, diffrec, startrec, endrec)
    | :math:`\phantom{xxx}` `|`-ctrl_init_ctrlvar('xx_atemp.effective.0000000001, 'c','xy')
    | :math:`\phantom{xxxx}` `|`-ctrl_set_fname(xx_fname,fname)
    | :math:`\phantom{xxxxxxxxxx}`    --> fname(1:3)=[,ad,hn]xx_atemp.effective.0000000001
    | :math:`\phantom{xxxx}` `|`-ctrl_set_globfld_xy(fname(2)) (with yadprefix="ad")

>     | :math:`\phantom{xxxxx}` `|` - MDSWRITEFIELD(adxx_atemp.effective.0000000001) (14 rec)

Author: remove (14 rec) and replace with (size diffrec) (similar to what i suggested in the yellow box)

Reviewer 2: So unless you (or @mjlosch ) object I'm going to remove the blue box here entirely (since it is duplicate), ie the code admonition format approach. The main pro for the blue box is I can add varlinks and filelinks. But I think the yellow looks cleaner (and offers a scroll bar for long lines), compact, ergo my preference.

Reviewer 2: actually, @jahn just discovered there is a parsed-literal block option that allows for links and text formatting. Seems the best of both worlds (although a white, not yellow background). Let me push this form and you can tell me what you think. I can bold comments now too. Also, I could also substitute an italic iter_no or something for 000000001 (although I added the suggestion to the first line, so could keep the 000000001).

Next PR, in this section I might suggest we not use the $ctrlvar format but «ctrlvar» which I've used elsewhere in the manual for something to be replace by the user (or a user choice). ie not clear what characters after the $ are part of the substitution. Also, there is potential confusion with optimization iteration and nIter0 /forward iteration that could be improved upon (again, these all are issues inherited from the original doc)

Author: yes, i know you said ignore the blue, so i was only looking at yellow anyway.. i was only commenting on blue because it has that extra "(14 rec)", which should now correspond to my suggested (size diffrec).

Aside from that, what you and Oliver found is also good, we can in the future use that rather than my hand-written format, which was also quite tedious.  I'll be looking out for your next edit in case i need to give inputs.
